src/extensions/default/JavaScriptRefactoring/RenameIdentifier.js

```diff
@@ -0,0 +1,149 @@
+    "use strict";
+
+    var EditorManager        = brackets.getModule("editor/EditorManager"),
+        ScopeManager         = brackets.getModule("JSUtils/ScopeManager"),
+        Session              = brackets.getModule("JSUtils/Session"),
+        MessageIds           = brackets.getModule("JSUtils/MessageIds"),
+        CommandManager       = brackets.getModule("command/CommandManager"),
+        Menus                = brackets.getModule("command/Menus"),
+        Strings              = brackets.getModule("strings");
+        //Commands
+    var refactorRename          = "javascript.renamereference";
+
+    var session             = null;  // object that encapsulates the current session state
+
+    //Create new session
+    function initializeSession(editor) {
+        session = new Session(editor);
+    }
+
+    //Post message to tern node domain that will request tern server to find refs
+    function getRefs(fileInfo, offset) {
+        ScopeManager.postMessage({
+            type: "getRefs",
+            fileInfo: fileInfo,
+            offset: offset
+        });
+
+        return ScopeManager.addPendingRequest(fileInfo.name, offset, MessageIds.TERN_REFS);
+    }
+
+    //Create info required to find reference
+    function requestFindRefs(session, document, offset) {
+        var path    = document.file.fullPath,
+            fileInfo = {
+                type: MessageIds.TERN_FILE_INFO_TYPE_FULL,
+                name: path,
+                offsetLines: 0,
+                text: ScopeManager.filterText(session.getJavascriptText())
+            };
+
+        var ternPromise = getRefs(fileInfo, offset);
+
+        return {promise: ternPromise};
+    }
+
+    //Do rename of identifier which is at cursor
+    function handleRename() {
+        var editor = EditorManager.getActiveEditor(),
+            offset, handleFindRefs;
+
+        if (editor.getSelections().length > 1) {
+            editor.displayErrorMessageAtCursor(Strings.ERROR_RENAME_MULTICURSOR);
+            return;
+        }
+        initializeSession(editor);
+
+
+        if (!editor || editor.getModeForSelection() !== "javascript") {
+            return;
+        }
+
+        var result = new $.Deferred();
+
+        /**
+         * Make a find ref request.
+         * @param {Session} session - the session
+         * @param {number} offset - the offset of where to jump from
+         */
+        function requestFindReferences(session, offset) {
+            var response = requestFindRefs(session, session.editor.document, offset);
+
+            if (response || response.hasOwnProperty("promise")) {
+                response.promise.done(handleFindRefs).fail(function () {
+                    result.reject();
+                });
+            }
+        }
+
+        /**
+         * Check if references are in this file only
+         * If yes then select all references
+         */
+        handleFindRefs = function (refsResp) {
```

Comment:
Maybe cleaner if this one was assigned inline above `requestFindReferences`?

`function handleFindRefs(refsResp)...`